components/fileqcs/fileQcsController.js

```diff
@@ -1,220 +1,519 @@
+    logger.error(e);
+    next(generateError(500, defaultMessage));
+  }
+}
+
 /** returns an object { validated: {}, errors: [] } */
-function validateObjectsFromUser(unvalidatedObjects, unvalidatedProject) {
+function validateObjectsFromUser (unvalidatedObjects) {
   let validationErrors = [];
-  let validatedParams = unvalidatedObjects.map(validateFileQcObject);
-  return { validated: validatedParams, errors: validationErrors };
-  
-  function validateFileQcObject(unvalidated) {
-    // project may be passed in separately from the fileqcs, or as part of the fileqcs array
-    const proj = unvalidatedProject || unvalidated.project;
-    let validated = {};
+  let validatedParams = unvalidatedObjects.map(unvalidated => {
     try {
-      validated.filepath = validateFilepath(unvalidated.filepath);
-      validated.qcpassed = validateQcStatus(unvalidated.qcstatus);
-      validated.username = validateUsername(unvalidated.username);
-      validated.comment = validateComment(unvalidated.comment);
-      validated.fileswid = validateSwid(unvalidated.fileswid);
-      validated.project = validateProject(proj);
+      return {
+        filepath: validateFilepath(unvalidated.filepath),
+        qcpassed: validateQcStatus(unvalidated.qcstatus, true),
+        username: validateUsername(unvalidated.username),
+        comment: validateComment(unvalidated.comment),
+        fileswid: validateSwid(unvalidated.fileswid),
+        project: validateProject(unvalidated.project)
+      };
     } catch (e) {
       if (e instanceof ValidationError) {
         validationErrors.push({ fileswid: unvalidated.fileswid, error: e.message });
         return null;
       } else {
-        console.log(e.error);
         throw e;
       } 
     }
-    return validated;
+  });
+
+  return { validated: validatedParams, errors: validationErrors };
+}
+
+
+/** success returns a single File Provenance Report result */
+function getSingleFprResult (swid) {
+  return new Promise((resolve, reject) => {
+    fpr.get('SELECT * FROM fpr WHERE fileswid = ?', [swid], (err, data) => {
+      if (err) reject(generateError(500, err));
+      resolve((data ? data : {}));
+    });
+  });
+}
+
+/** success returns a single FileQC result */
+function getSingleFqcResult (swid) {
+  return new Promise((resolve, reject) => {
+    const sql = 'SELECT * FROM FileQc WHERE fileswid = $1';
+    pg.any(sql, [swid])
+      .then(data => { 
+        if (!data || data.length == 0) resolve({ fileqc: {}, errors: [] });
+        resolve({ fileqc: data[0], errors: [] });
+      })
+      .catch(err => {
+        logger.error(err);
+        reject(generateError(500, 'Error retrieving record'));
+      });
+  });
+}
+
+/** success returns an array of File Provenance results filtered by the given project */
+function getFprResultsByProject (project) {
+  return new Promise((resolve, reject) => {
+    fpr.all('SELECT * FROM fpr WHERE project = ? ORDER BY fileswid ASC', [project], (err, data) => {
```

Comment:
Some of these `?` and some are `$1` and some are `${x}`. Is there a reason?

Reply:
`?` is for SQLite, and the `$` ones are Postgres. I've used `${x}` where there are three or more arguments (and in these cases, the arguments are coming into the function as an object already), whereas `$1` is used when working with just a couple of parameters. If you think it's needlessly complicated, I'll use the `${x}` syntax for all Postgres statements and create objects for the 1-2 parameter statements.

Reply:
I've added a comment to the file which indicates the conventions I've used for parameterization.